torchtune/utils/ppo/_generation.py

```diff
@@ -0,0 +1,185 @@
+        max_generated_tokens (int): number of tokens to be generated
+        pad_id (int): token ID to use for padding, default 0.
+        temperature (float): value to scale the predicted logits by, default 1.0.
+        top_k (Optional[int]): If specified, we prune the sampling to only token ids within the top_k probabilities,
+            default None.
+        stop_tokens (Optional[List[int]]): If specified, generation is stopped when any of these tokens are generated,
+            default None.
+        custom_generate_next_token (Optional[Callable]): If specified, we'll use the ``custom_generate_next_token function``.
+            This is generally only useful if you want to specify a ``torch.compile`` version of the generate next token for
+            performance reasons. If None, we use the default ``generate_next_token`` function. Default is None.
+
+    Examples:
+        >>> model = torchtune.models.llama3.llama3_8b()
+        >>> tokenizer = torchtune.models.llama3.llama3_tokenizer()
+        >>> prompt = [0, 0, 0] + tokenizer("Hi my name is") # substitute 0 with pad_id
+        >>> output = generate(model, torch.tensor(prompt), max_generated_tokens=100)
+        >>> print(tokenizer.decode(output[0]))
+        ?? ?? ?? Hi my name is Jeremy and I'm a friendly language model assistant!
+
+    Returns:
+        List[List[int]]: collection of lists of generated tokens
+    """
+    prompt = prompt.view(1, -1) if prompt.ndim == 1 else prompt
+
+    bsz, prompt_length = prompt.size()
+    generated_tokens = prompt.clone()
+    generated_tokens[generated_tokens == pad_id] = 0
+
+    if stop_tokens is not None:
+        # convert stop tokens to tensor for easy matching
+        stop_tokens = (
+            torch.tensor(stop_tokens, device=prompt.device) if stop_tokens else None
+        )
+        # keeps track at a high level if we've already hit a stop token in a sequence so we can early stop
+        stop_token_reached = torch.zeros(bsz, dtype=torch.bool, device=prompt.device)
+        # everything in stop_token_mask starts as 1s, and we'll set them to 0 for sequences
+        # that already hit a stop token
+        stop_token_mask = torch.ones(
+            (bsz, prompt_length), dtype=torch.int32, device=prompt.device
+        )
+
+    if custom_generate_next_token is None:
+        custom_generate_next_token = generate_next_token
+
+    for i in range(max_generated_tokens):
+        # update stop_token_mask if we reached a stop token in a previous step
+        # by appending the logical not of stop_token_reached to the end of the mask
+        # reshaped to be bsz first
+        if stop_tokens is not None:
+            stop_token_mask = torch.cat(
+                [stop_token_mask, ~stop_token_reached.reshape(bsz, 1)], dim=-1
+            )
+
+        padding_mask = generated_tokens == pad_id
+        mask = get_causal_mask(generated_tokens, padding_id=pad_id, dtype=dtype)
```

Comment:
I can probably be more efficient about this, we can update position IDs and attention masks by just adding the next position id per batch, and the next column+row to the attention mask, without recalculating from scratch.

Reply:
adding to this, I can also just return the calculated input ids and masks here rather than recalculate them in the recipe
 - or even pre-calculate them before generation, since we know what the masks and pos ids for the responses will be (sequential from the pos id of the last token in the sequence, and no tokens masked out for attention)